R/pkg/R/DataFrame.R

```diff
@@ -2199,3 +2199,97 @@ setMethod("coltypes",
 
             rTypes
           })
+
+#' Display the structure of a DataFrame, including column names, column types, as well as a
+#' a small sample of rows.
+#' @name str
+#' @title Compactly display the structure of a dataset
+#' @rdname str
+#' @family DataFrame functions
+#' @param object a DataFrame
+#' @examples \dontrun{
+#' # Create a DataFrame from the Iris dataset
+#' irisDF <- createDataFrame(sqlContext, iris)
+#' 
+#' # Show the structure of the DataFrame
+#' str(irisDF)
+#' }
+setMethod("str",
+          signature(object = "DataFrame"),
+          function(object) {
+
+            # TODO: These could be made global parameters, though in R it's not the case
+            MAX_CHAR_PER_ROW <- 120
+            MAX_COLS <- 100
+
+            # Get the column names and types of the DataFrame
+            names <- names(object)
+            types <- coltypes(object)
+
+            # Get the number of rows.
+            # TODO: Ideally, this should be cached
+            cachedCount <- nrow(object)
+
+            # Get the first elements of the dataset. Limit number of columns accordingly
+            dataFrame <- if (ncol(object) > MAX_COLS) {
+                           head(object[, c(1:MAX_COLS)])
+                         } else {
+                           head(object)
+                         }
+
+            # The number of observations will be displayed only if the number
+            # of rows of the dataset has already been cached.
+            if (!is.null(cachedCount)) {
```

Comment:
If we need cache for nrow, it should be done internally in nrow(), not done in consumer side of nrow.

@rxin, I don't understand the meaning of "the query plan can be nondeterministic ". Since DataFrame is immutable, why it is not safe to cache the number of rows? No matter of the query plan, the result should be same, right?